### DISCUSSION TITLE
How to convert map type to object

### DISCUSSION THREAD

**Discussion by rostik404** [grpc/grpc-web#1422] [Q&A - General]
Hello!
I have a next proto type:
```
message Test {
    map <string, string> values = 1;
}

```
Next, I'm making an object of this type:
```
const test = new Test().getValuesMap().set("test", "1")
```
And then, I'm trying to lead this type to an object:

```
console.log(JSON.stringify(test.toObject(), null, 4));
```
So, I'm getting this structure:
```
{
    valuesMap: [
        ["test", "1"]
    ]
}
```
But I'm expecting a structure like this:
```
{
    values: {
        test: "1"
    }
}
```

According to [this](https://github.com/protocolbuffers/protobuf/blob/2b7430d96aeff2bb624c8d52182ff5e4b9f7f18a/js/maps_test.js#L126) file, the behaviour with "valuesMap" is valid. However, is there any way to convert a proto to an object as I'm expecting?

Also, the provided example is synthetic. In real case I have a recursive structure like this:
```
{
    "root": {
        "valuesMap": [
            [
                "key1",
                {
                    "name": "some-name"
                    "nested": {
                        "valuesMap": [
                            [
                                "key2",
                                {
                                    "name": "other-name"
...

``` 
So instead of using just `res.toObject()`, I need to write some recursive function on a client to transform this object.

**Comment by sampajano** [comment]
Hi @rostik404 

It's more of a protobuf question, but thanks for asking here. :)

You can maybe take a look at the map implementation / APIs to see if there's anything you can do better:
https://github.com/protocolbuffers/protobuf-javascript/blob/main/map.js

From what I can roughly tell form a quick look, it seems that Map could e implemented as array internally, and hence maybe that's why "toObject" returned as such.

Not sure if there's other quick JS tricks you could do to quickly do such conversions.
